ambari-server/src/main/java/org/apache/ambari/server/agent/stomp/AgentHostDataHolder.java

```diff
@@ -55,11 +55,19 @@ public T getUpdateIfChanged(String agentHash, Long hostId) throws AmbariExceptio
   public T initializeDataIfNeeded(Long hostId, boolean regenerateHash) throws AmbariException {
     T hostData = data.get(hostId);
     if (hostData == null) {
-      hostData = getCurrentData(hostId);
-      if (regenerateHash) {
-        regenerateDataIdentifiers(hostData);
+      updateLock.lock();
```

Comment:
Only a write lock without any read lock, as a scheme will work here? Can you please add some comment as to why?

Reply:
This part will be changed in separate patch.
I'm going to get rid of locking here in favor of compute* methods of concurrent hashmap.
Requires some additional changes related to handling updates, so was not done here.

Reply:
Ok, that would be better for sure!